File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
# 강사님 풀이

n,k=map(int,input().split())
knapsack=[[0 for _ in range(k+1)] for _ in range(n+1)]

item=[[0,0]]
for i in range(1,n+1):  # 아이템 입력
    item.append(list(map(int,input().split())))

for i in range(1,n+1):  # 아이템의 개수 만큼 반복
    for j in range(1,k+1): # 최대 무게까지 반복
        w=item[i][0]
        value=item[i][1]
        if j<w: # 가방에 넣을 수 없다면..
            knapsack[i][j]=knapsack[i-1][j]
        else: # 가방에 넣을 수 있다면..
            knapsack[i][j]=max(knapsack[i-1][j],value+knapsack[i-1][j-w])
                               # 위에값     vs   현재아이템 가치+전 단계에서 구한 남은 무게 가치
print(knapsack[n][k])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int[][] knapsack = new int[n + 1][k + 1];

        int[][] item = new int[n][2];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            item[i][0] = Integer.parseInt(st.nextToken());
            item[i][1] = Integer.parseInt(st.nextToken());
        }

        for (int i = 1; i < n + 1; i++) {
            for (int j = 0; j < k + 1; j++) {
                int w = item[i - 1][0];
                int value = item[i - 1][1];
                if (j < w) {
                    knapsack[i][j] = knapsack[i - 1][j];
                } else {
                    knapsack[i][j] = Math.max(knapsack[i - 1][j], value + knapsack[i - 1][j - w]);
                }
            }
        }
        System.out.println(knapsack[n][k]);
    }
}